### PR TITLE
support extended capabilities

### PR DESCRIPTION
**Pull request by jlevon**

### CODE REVIEW

**Review by swapnili — COMMENTED** [review]
It would be nice to make sample/lspci or client/server to use extended caps.

File: lib/pci_caps.c [hunk 1]
```diff
@@ -432,79 +551,154 @@ vfu_pci_add_capability(vfu_ctx_t *vfu_ctx, size_t pos, int flags, void *data)
...
+        errno = EINVAL;
+        return 0;
+    }
+
+    if (offset == 0) {
+        offset = PCI_CFG_SPACE_SIZE;
+        hdr = (void *)pci_config_space_ptr(vfu_ctx, offset);
+    } else {
+        hdr = (void *)pci_config_space_ptr(vfu_ctx, offset);
+        hdr = (void *)pci_config_space_ptr(vfu_ctx, hdr->next);
```
[the remark this answers is not in the document]
Comment (jlevon) [reply lib/pci_caps.c:660]: If offset is set, it's the offset of the capability that was last returned. So we need to look at that header, find its next field, and pick that up.
e.g.

offset = pci_find_ext_capaibility(CAP_VNDR); // returns 256 for the first cap, that is offset is now 256
offset = pci_find_next_ext_capability(offset, CAP_VNDR); // we pass in 256, and look at its ->next field to find our next cap offset, let's say 512
offset = pci_find_next_ext_capability(offset, CAP_VNDR); // this time we look at the cap at 512, its ->next field is zero, we return 0 and we're done.

**Review by jlevon — COMMENTED** [review]
> client/server to use extended caps

good idea.

File: lib/pci_caps.c [hunk 2]
```diff
@@ -282,6 +307,22 @@ cap_write_vendor(vfu_ctx_t *vfu_ctx, struct pci_cap *cap UNUSED, char *buf,
     return count;
 }
 
+static ssize_t
+ext_cap_write_dsn(vfu_ctx_t *vfu_ctx, struct pci_cap *cap, char *buf UNUSED,
+                  size_t count UNUSED, loff_t offset UNUSED)
+{
+    vfu_log(vfu_ctx, LOG_ERR, "%s capability is read-only\n", cap->name);
+    return -EINVAL;
```
[the remark this answers is not in the document]
Comment (jlevon) [reply lib/pci_caps.c:315]: sure

File: lib/pci_caps.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (swapnili) [reply lib/pci_caps.c:660]: Ahh yes. I dint saw that this function is public and can be used to traverse multiple caps with same  id.

**Review by swapnili — APPROVED** [review]

**Review by swapnili — APPROVED** [review]

**Pull request merged**